runtime/j9vm/javanextvmi.cpp

```diff
@@ -253,20 +253,32 @@ JVM_VirtualThreadMountBegin(JNIEnv *env, jobject thread, jboolean firstMount)
 	J9VMThread *currentThread = (J9VMThread *)env;
 	J9JavaVM *vm = currentThread->javaVM;
 	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	BOOLEAN releasedLiveVirtualThreadListMutex = FALSE;
 
 	vmFuncs->internalEnterVMFromJNI(currentThread);
 	f_monitorEnter(vm->liveVirtualThreadListMutex);
 	j9object_t threadObj = J9_JNI_UNWRAP_REFERENCE(thread);
 
 	assert(IS_JAVA_LANG_VIRTUALTHREAD(currentThread, threadObj));
 
-	while (0 != J9OBJECT_I64_LOAD(currentThread, threadObj, vm->virtualThreadInspectorCountOffset)) {
+	while (vm->inspectingLiveVirtualThreadList
+	|| (0 != J9OBJECT_I64_LOAD(currentThread, threadObj, vm->virtualThreadInspectorCountOffset))
+	) {
 		/* Thread is being inspected or unmounted, wait. */
-		VM_VMHelpers::pushObjectInSpecialFrame(currentThread, threadObj);
 		vmFuncs->internalExitVMToJNI(currentThread);
 		f_monitorWait(vm->liveVirtualThreadListMutex);
+		if (OMR_ARE_ANY_BITS_SET(currentThread->publicFlags, J9_PUBLIC_FLAGS_HALT_THREAD_JAVA_SUSPEND)) {
+			/* Release the monitor before suspending. */
+			f_monitorExit(vm->liveVirtualThreadListMutex);
+			releasedLiveVirtualThreadListMutex = TRUE;
```

Comment:
Can you indicate the location where the monitor should be released and reacquired?

Reply:
VMAccess and monitor usage are interlinked since we are reading from an object and the object field is synchronized by the monitor. Not sure if we can efficiently achieve the above suggestion.

Reply:
I'll have to think about this a bit - intially I was thinking you just remove the "if suspended" check.

Reply:
I think the pattern should be:
```
enter VM
acquire monitor
while (condition) {
   exit VM
   wait
   release monitor
   enter VM
   acquire monitor
}
```

Reply:
Makes sense. This will protect us from misbehaviour in scenarios where `currentThread->publicFlags` is outdated. Trying it out.

Reply:
Updated. Local testing with OpenJDK tests (SuspendResume2 and SuspendResumeAll) passes.